### ISSUE TITLE
[Bug] Progress report don't show unfulfilled criteria.

### ISSUE THREAD

**Issue by ARGAMX** [CorsixTH/CorsixTH#2750]
### Describe the issue

If the player kills more than a certain percentage of patients, then in Progress report instead of showing required hospital cost, the death rate criterion is displayed. So death criteria overlaps hospital cost criteria.
Thus, the player does not see hospital cost criterion at all.
If player fulfills all the other 4 criteria, he will be in a situation where all the displayed criteria are fulfilled, but the level does not end.
And the player doesn't understand why.

<img width="329" alt="Screenshot 2024-12-26 at 21 31 13" src="https://github.com/user-attachments/assets/2403d1c8-4d93-46ea-b27e-a1f14be51d38" />

Example of such case: #2675

### Steps to Reproduce

1. *Load save game attached to #2675 ticket*
2. Observe that level can't be cleared despite fulfilling all visible criteria.

### Expected Behaviour

Expected Behaviour should resolve that issue.

As a variants:
1. Instead of showing fulfilled criteria, Progress report can show unfulfilled first. With next priorities: I. Critical criteria (in red zone), II. Unfulfilled criteria, III. Fulfilled criteria.
2. Scrolling like left and right arrows around the table. (@tobylane proposed)

Also take into account that there are currently 5 criteria for victory, but Since CorsixTH 0.69 and up are going to have 2 more win conditions.

### Have you enabled any current CorsixTH Experimental Features?

- [ ] Right mouse panning
- [ ] Falling actions

### Save Game

1

### CorsixTH Version

0.68.0

### Operating System

1

### Theme Hospital Version

CD

**Comment by Matroftt** [comment]
I'll send there image of scaling concept soon. Basically, when there is 5 or less criteria it shows them like usual, but when there is more, progress report shrinks them, so they all can appear on the screen

**Comment by lewri** [comment]
#2578 will be related to this, as it now only takes 5 criteria for display with that patch (though I don't think it currently affects the situation any more than before).

**Comment by tobylane** [comment]
Previously, the progress dialog excluded the percentage_killed criteria, as that would be the last of the six campaign goals looked at.

Now, it sorts by the gap from the measure to the target, ie the current hospital value to the win/lose value. It sorts the wrong way, it should be less than, not greater than. But this shouldn't account for the bug in the screenshot, as losing goals are considered before winning goals.

Comparing the gaps of different criteria is somewhat meaningless. Would a percentage progress be the right way to do this? The non-zero start attribute of the balance and reputation criteria tables enables measuring 600 on the way to a goal of 700 as 50%, not 85%. This percentage wouldn't be shown to the user, it's an internal value like the height of the graph columns.

https://github.com/CorsixTH/CorsixTH/blob/c0c9d26d79c6b319dbe0a34d29a16f6d446f0521/CorsixTH/Lua/endconditions.lua#L154-L158

**Comment by Alberth289346** [comment]
> Would a percentage progress be the right way to do this?

That sounds better than the values themselves at least, due to scaling. If I can kill 5 patients and have killed 4 already, I'd want to consider that criteria more important than the 600 to 700 reputation.

In a sense, you want something like "how easy is it to exceed the limit". Without taking time into account, a percentage is likely enough.  
If you do include time, you get a notion  "most likely to exceed first" (ie rate of growth). The problem becomes then finding a good prediction of the rate. Not sure that exists.

**Comment by ARGAMX** [comment]
Still, the Progress Percentage won't solve the topic problem, right?

**Comment by tobylane** [comment]
No, it won't. Lose goals where the criteria are over their boundary are considered more important than any win goals. It relies on the level writer choosing the right information to show the player (recently changed in that campaign, below). That assumption may have to stay until the dialog UI is changed in the way you or Matroftt suggest. I'm wary of hiding completed goals, but shifting is better. The mix of lose and win goals are ordered by number to deal with a UI issue and to match TH expectations, which puts CorsixTH's new criteria at the end. That could be redone too.

The background is loaded with gfx:loadRaw, which I know nothing about. How easy would it be to widen the chart by repeating the background for more columns, pushing the left half of the image leftwards?

https://github.com/CorsixTH/CorsixTH/blob/fa8437a891b6127b1b04b11dc2972baf1cad7498/CorsixTH/Campaigns/fourcorners.level#L214

**Comment by ARGAMX** [comment]
It might be possible to fix this together with #2511

**Comment by Alberth289346** [comment]
> The background is loaded with gfx:loadRaw, which I know nothing about. How easy would it be to widen the chart by repeating the middle 30 pixels of the background for more columns, pushing the left half of the image leftwards?

It seems to load one image of 64x480. 1 pixel is 1 byte, so it loads 640*480 bytes.
Loading seems to end in `app.lua`, line 1543 `readBitmapDataFile`, which just reads the 640*480 bytes, and possibly decompresses if RNC compression was applied.

If you want to change the picture, you have to copy some bytes to a different index. The file is most likely row-oriented, thus 640 bytes for the first row, next 640 for the second row, and so on.
pixel `(x, y)` is at index `x + y * 640` (with `x`, `y`, and the `index` all starting at `0`). Pixel `(0, 0)` is thus at index `0` at the top-left of the image.

Simplest to try how it looks is by taking a screenshot, and then move the desired part in a graphics editor. You can see how it looks then.

**Comment by ARGAMX** [comment]
> The background is loaded with gfx:loadRaw, which I know nothing about. How easy would it be to widen the chart by repeating the middle 30 pixels of the background for more columns, pushing the left half of the image leftwards?

> If you want to change the picture, you have to copy some bytes to a different index. The file is most likely row-oriented, thus 640 bytes for the first row, next 640 for the second row, and so on.
> pixel `(x, y)` is at index `x + y * 640` (with `x`, `y`, and the `index` all starting at `0`). Pixel `(0, 0)` is thus at index `0` at the top-left of the image.

Wouldn't it be 10 times easier to edit the image in a graphics editor and store this new image it in our project as a image asset?

I understand what Albert wrote. But to me it sounds like creating a Frankenstein instead of once edit the image in the editor and keep it.

**Comment by Matroftt** [comment]
> I'll send there image of scaling concept soon. Basically, when there is 5 or less criteria it shows them like usual, but when there is more, progress report shrinks them, so they all can appear on the screen

A little late
Basically, if there is 5 criteria or less, progress table should leave their position as is. But if there is more than 5 of them, it should move criterias so they all can fit. (Probably, it should shrink bars too)

![Image](https://github.com/user-attachments/assets/ccc414c1-0990-431b-b160-405ed67b8a13)

**Comment by Alberth289346** [comment]
> Wouldn't it be 10 times easier to edit the original image in a graphics editor and store this new image it in our project as a image asset?

Technically yes, but copyright says no. We are not allowed to distribute copyrighted graphics.

In other words, it would work iff every user prepares such an image for themselves.

**Comment by TheCycoONE** [comment]
What if instead of shrinking the columns we provided 90s style left/right buttons to shift through the criteria

**Comment by TheCycoONE** [comment]
<img width="652" height="496" alt="Image" src="https://github.com/user-attachments/assets/2dbad18a-7f1d-499b-afc4-d42e279f38ab" /> - my bad artistic rendering, but there may be better placement options.

**Comment by Alberth289346** [comment]
> What if instead of shrinking the columns we provided 90s style left/right buttons to shift through the criteria

If you insert a short slider between the arrows you can visualize what part of the criteria it being displayed.

**Comment by ARGAMX** [comment]
This issue is still relevant in CorsixTH 0.70.1.
